### ISSUE TITLE
yo angular-fullstack needs node v0.12

### ISSUE THREAD

**Issue by m66n** [FreeCodeCamp/FreeCodeCamp#2722]
Challenge http://freecodecamp.com/challenges/waypoint-get-set-for-basejumps has an issue. Please describe how to reproduce it, and include links to screenshots if possible.

The angular-fullstack generator fails with node 0.10.35 (the default version on Cloud 9). This is the error:

```
events.js:72
        throw er; // Unhandled 'error' event
              ^
TypeError: Object #<Object> has no method 'isAbsolute'
    at destinationPath (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/util.js:80:13)
    at Object.relativeRequire (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/util.js:88:8)
    at Generator.<anonymous> (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/script-base.js:40:25)
    at eval (eval at template (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/node_modules/yeoman-generator/node_modules/lodash/dist/lodash.js:6306:22), <anonymous>:8:11)
    at underscore (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/node_modules/yeoman-generator/lib/util/engines.js:23:45)
    at Generator.actions.engine (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/node_modules/yeoman-generator/lib/actions/actions.js:216:32)
    at Generator.template (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/node_modules/yeoman-generator/lib/actions/actions.js:191:19)
    at /home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/util.js:158:14
    at Array.forEach (native)
    at Object.processDirectory (/home/ubuntu/.nvm/v0.10.35/lib/node_modules/generator-angular-fullstack/util.js:128:9)
```

I used nvm to install node 0.12.7 and modified the script in step 11 to get around it. Also, the generator asks an additional question -- `What would you like to write tests with?` -- that is not in the list of checkable items for the waypoint.

**Comment by mizujin** [comment]
The first step also needs to be modified :

rm -rf * && echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/v0.10.35/lib/node_modules" >> ~/.bashrc && source ~/.bashrc && npm install -g yo grunt grunt-cli generator-angular-fullstack && yo angular-fullstack

Should now be :

rm -rf * && echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/**v0.12**/lib/node_modules" >> ~/.bashrc && source ~/.bashrc && npm install -g yo grunt grunt-cli generator-angular-fullstack && yo angular-fullstack

**Comment by rtperson** [comment]
I just worked on this tonight with a few FCC'ers. I should have a pull request ready soon. (Essentially all it takes is changing the big BASH command to include a call to nvm and a change the export to .bashrc)

**Comment by rtperson** [comment]
Like so:

    rm -rf * && nvm install v0.12.7 && echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/v0.12.7/lib/node_modules" >> ~/.bashrc && source ~/.bashrc && npm install -g yo grunt grunt-cli generator-angular-fullstack && yo angular-fullstack

**Comment by m66n** [comment]
I get this on a fresh Cloud 9 workspace:

```
~/workspace $ rm -rf * && nvm install v0.12.7 && echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/v0.12.7/lib/node_modules" >> ~/.bashrc && source ~/.bashrc && npm install -g yo grunt grunt-cli generator-angular-fullstack && yo angular-fullstack
######################################################################## 100.0%
Now using node v0.12.7
npm ERR! Linux 3.14.13-c9
npm ERR! argv "/home/ubuntu/.nvm/versions/node/v0.12.7/bin/node" "/home/ubuntu/.nvm/versions/node/v0.12.7/bin/npm" "install" "-g" "yo" "grunt" "grunt-cli" "generator-angular-fullstack"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3
npm ERR! code ECONNRESET
npm ERR! errno ECONNRESET
npm ERR! syscall read

npm ERR! network read ECONNRESET
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network 
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
/home/ubuntu/.nvm/versions/node/v0.12.7/bin/grunt -> /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/grunt-cli/bin/grunt

> spawn-sync@1.0.13 postinstall /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/yo/node_modules/cross-spawn/node_modules/spawn-sync
> node postinstall

/home/ubuntu/.nvm/versions/node/v0.12.7/bin/yo -> /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/yo/lib/cli.js

> yo@1.4.7 postinstall /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/yo
> yodoctor


Yeoman Doctor
Running sanity checks on your system

✔ Global configuration file is valid
✔ Node.js version
✔ No .bowerrc file in home directory
✔ No .yo-rc.json file in home directory
✔ npm version
✖ NODE_PATH matches the npm root
npm global root value is not in your NODE_PATH

[Info]
  NODE_PATH = /mnt/shared/lib/node_modules:/home/ubuntu/.nvm/v0.12.7/lib/node_modules
  npm root  = /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules

[Fix] Append the npm root value to your NODE_PATH variable
Add this line to your .bashrc
    export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules
  Or run this command
    echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules" >> ~/.bashrc && source ~/.bashrc


Found potential issues on your machine :(
npm WARN deprecated CSSselect@0.4.1: the module is now available as 'css-select'
npm WARN deprecated CSSwhat@0.4.7: the module is now available as 'css-what'
grunt-cli@0.1.13 /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/grunt-cli
├── resolve@0.3.1
├── nopt@1.0.10 (abbrev@1.0.7)
└── findup-sync@0.1.3 (lodash@2.4.2, glob@3.2.11)

yo@1.4.7 /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/yo
├── figures@1.3.5
├── titleize@1.0.0
├── array-uniq@1.0.2
├── user-home@1.1.1
├── async@1.4.2
├── opn@1.0.2
├── humanize-string@1.0.1 (decamelize@1.0.0)
├── yeoman-character@1.0.1 (supports-color@1.3.1)
├── findup@0.1.5 (commander@2.1.0, colors@0.6.2)
├── string-length@1.0.1 (strip-ansi@3.0.0)
├── chalk@1.1.1 (escape-string-regexp@1.0.3, supports-color@2.0.0, ansi-styles@2.1.0, strip-ansi@3.0.0, has-ansi@2.0.0)
├── meow@3.3.0 (object-assign@3.0.0, indent-string@1.2.2, camelcase-keys@1.0.0, minimist@1.2.0)
├── repeating@1.1.3 (is-finite@1.0.1)
├── root-check@1.0.0 (sudo-block@1.2.0, downgrade-root@1.1.0)
├── sort-on@1.2.2 (arrify@1.0.0, dot-prop@2.2.0)
├── configstore@1.2.1 (os-tmpdir@1.0.1, object-assign@3.0.0, graceful-fs@4.1.2, uuid@2.0.1, xdg-basedir@2.0.0, osenv@0.1.3, write-file-atomic@1.1.2, mkdirp@0.5.1)
├── package-json@1.2.0 (registry-url@3.0.3)
├── npm-keyword@1.2.0 (registry-url@3.0.3)
├── update-notifier@0.5.0 (is-npm@1.0.0, latest-version@1.0.1, semver-diff@2.0.0)
├── yosay@1.0.5 (ansi-regex@1.1.1, strip-ansi@2.0.1, ansi-styles@2.1.0, word-wrap@1.1.0, pad-component@0.0.1, taketalk@1.0.0, minimist@1.2.0, string-width@1.0.1)
├── fullname@1.1.1 (npmconf@2.1.2)
├── got@3.3.1 (lowercase-keys@1.0.0, timed-out@2.0.0, is-stream@1.0.1, is-redirect@1.0.0, prepend-http@1.0.2, object-assign@3.0.0, infinity-agent@2.0.3, nested-error-stacks@1.0.1, read-all-stream@3.0.1, duplexify@3.4.2)
├── cross-spawn@0.4.1 (lru-cache@2.6.5, spawn-sync@1.0.13)
├── yeoman-environment@1.2.7 (escape-string-regexp@1.0.3, log-symbols@1.0.2, diff@1.4.0, text-table@0.2.0, untildify@2.1.0, debug@2.2.0, mem-fs@1.1.0, globby@2.1.0, grouped-queue@0.3.0)
├── lodash@3.10.1
├── yeoman-doctor@1.4.0 (log-symbols@1.0.2, object-values@1.0.0, each-async@1.1.1, semver@4.3.6, bin-version-check@2.1.0, twig@0.8.2)
├── inquirer@0.8.5 (ansi-regex@1.1.1, cli-width@1.0.1, through@2.3.8, readline2@0.1.1, rx@2.5.3)
└── insight@0.6.0 (object-assign@2.1.1, async@0.9.2, lodash.debounce@3.1.1, os-name@1.0.3, tough-cookie@1.2.0, request@2.61.0)

grunt@0.4.5 /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/grunt
├── which@1.0.9
├── dateformat@1.0.2-1.2.3
├── eventemitter2@0.4.14
├── getobject@0.1.0
├── rimraf@2.2.8
├── colors@0.6.2
├── async@0.1.22
├── hooker@0.2.3
├── grunt-legacy-util@0.2.0
├── exit@0.1.2
├── nopt@1.0.10 (abbrev@1.0.7)
├── minimatch@0.2.14 (sigmund@1.0.1, lru-cache@2.6.5)
├── glob@3.1.21 (inherits@1.0.1, graceful-fs@1.2.3)
├── lodash@0.9.2
├── coffee-script@1.3.3
├── underscore.string@2.2.1
├── iconv-lite@0.2.11
├── findup-sync@0.1.3 (glob@3.2.11, lodash@2.4.2)
├── js-yaml@2.0.5 (argparse@0.1.16, esprima@1.0.4)
└── grunt-legacy-log@0.1.2 (grunt-legacy-log-utils@0.1.1, underscore.string@2.3.3, lodash@2.4.2)

generator-angular-fullstack@2.1.1 /home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/generator-angular-fullstack
├── chalk@0.4.0 (has-color@0.1.7, ansi-styles@1.0.0, strip-ansi@0.1.1)
├── generator-ng-component@0.0.10
├── wiredep@0.4.2 (chalk@0.1.1, lodash@1.3.1)
└── yeoman-generator@0.17.7 (dargs@2.1.0, isbinaryfile@2.0.4, class-extend@0.1.1, diff@1.4.0, text-table@0.2.0, underscore.string@2.4.0, mime@1.3.4, async@0.9.2, grouped-queue@0.3.0, nopt@3.0.3, cross-spawn@0.2.9, chalk@0.5.1, debug@1.0.4, run-async@0.1.0, mkdirp@0.5.1, shelljs@0.3.0, iconv-lite@0.4.11, glob@4.5.3, file-utils@0.2.2, rimraf@2.4.3, findup-sync@0.1.3, github-username@1.1.1, download@1.0.7, lodash@2.4.2, cheerio@0.17.0, gruntfile-editor@0.2.0, request@2.61.0, inquirer@0.7.3)
npm ERR! Linux 3.14.13-c9
npm ERR! argv "/home/ubuntu/.nvm/versions/node/v0.12.7/bin/node" "/home/ubuntu/.nvm/versions/node/v0.12.7/bin/npm" "install" "-g" "yo" "grunt" "grunt-cli" "generator-angular-fullstack"
npm ERR! node v0.12.7
npm ERR! npm  v2.11.3

npm ERR! Callback called more than once.
npm ERR! 
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /home/ubuntu/workspace/npm-debug.log
```

**Comment by rtperson** [comment]
I'm seeing the same thing. I was able to get everything working by separating out the call to add the NODE_PATH to .bashrc into its own line. Not sure why it's not able to run within a one liner. I'm going to have to come back to this one tomorrow.

**Comment by rtperson** [comment]
It looks like npm is picky about how NODE_PATH is set, and using nvm has changed where the NODE_PATH lives. I fixed it with the following command:

    rm -rf * && nvm install v0.12.7 && echo "export NODE_PATH=/home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules" >> ~/.bashrc && source ~/.bashrc && npm install -g yo grunt grunt-cli generator-angular-fullstack && yo angular-fullstack

I received out-of-memory errors every time I ran this, but I fixed this by running `yo angular-fullstack`.

However, having done all this, I ran into a host of dependency programs on later steps of the waypoint. `npm list` shows a lot of unmet dependencies, and `grunt serve` fails with the following message:

```
rtperson@test-of-waypoint:~/workspace $ grunt serve
/home/ubuntu/workspace/node_modules/grunt/node_modules/findup-sync/lib/findup-sync.js:33
    }).flatten().uniq().value();
       ^
TypeError: undefined is not a function
    at Object.module.exports [as findup] (/home/ubuntu/workspace/node_modules/grunt/node_modules/findup-sync/lib/findup-sync.js:33:8)
    at Task.task.init (/home/ubuntu/workspace/node_modules/grunt/lib/grunt/task.js:427:16)
    at Object.grunt.tasks (/home/ubuntu/workspace/node_modules/grunt/lib/grunt.js:120:8)
    at Object.module.exports [as cli] (/home/ubuntu/workspace/node_modules/grunt/lib/grunt/cli.js:38:9)
    at Object.<anonymous> (/home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules/grunt-cli/bin/grunt:45:20)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10) 
```

So, to recap, we can't run the waypoint with the default node, and apparently upgrading to the most recent stable version of node causes problems with grunt (if not with the angular-fullstack generator itself). Right now I'm at a loss as to how to fix this.

**Comment by Skmsoumya** [comment]
Hey guys after some head banging I fount an fix to the problem. Follow the following process and it will be solved. 

* First clean up the directory using `rm -rf *`

* Then update node by running the following commands

```
nvm install 0.12
nvm use 0.12
nvm alias default v0.12
```

* Then check the version of npm by `npm -v`. It should return - v0.12.7.

* Now install yo (do not try to install yo, bower and grunt together in a single command line)

```
npm install -g yo
```

   Now you will see a that it is installing yo but yeoman doctor will report something like
 
```
npm global root value is not in your NODE_PATH
```

Just below it you will find a info tag and a fix tag, where it will tell you to run a command like 

```
echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules" >> ~/.bashrc && source ~/.bashrc
```

Copy that command and run it. 

* Now install bower by running

```
npm install -g bower
```

* Then install Grunt by running

```
npm install -g grunt-cli
```

* Well we are almost finished. Now you just have to install the generator by running 

```
npm install -g generator-angular-fullstack
```

* And then scaffold your web app by running

```
yo angular-fullstack
```

* Well now you can follow along the tutorial. It will run fine :-)

**Comment by m66n** [comment]
I also upgraded `npm` using `npm -g install npm` to get version `2.14.0`.

But I still run into the memory limit hit that rtperson mentions. I run `bower install & npm install` twice and it seems to complete, but `grunt serve` fails with an `Unexpected token ILLEGAL` in `lodash.js:1839`. When I run these steps on a Linux PC, there seem to be a lot of git fetches and builds that are not occurring on the Cloud 9 VM.

**Comment by mizujin** [comment]
The same occurs to me to. I did update node to 0.12.7 and successfully did yo angular-fullstack. But when I try to start the server with grunt serve I get all kind of errors with various package, starting with esprima.

**Comment by Skmsoumya** [comment]
Before running the tutorial I just updated Node. After that when I checked the version of  npm it shows the latest one. So I think installing node itself updates npm and you don't have to do it.

**Comment by gwenf** [comment]
@Skmsoumya's walkthrough (earlier in this thread) worked for me. I followed it exactly and then started Quincy's tutorial at 9 minutes in where he explains what to do with the red error notices that pop up.

**Comment by m66n** [comment]
I got this to work with no memory limit crashes by using older versions of nodejs and generator-angular-fullstack:
1. Clean up: `rm -rf *`
2. Get latest nodejs v0.10 `nvm install 0.10` --> `Now using node v0.10.40`
3. Get latest npm: `npm install -g npm` --> `npm@2.14.1`
4. Install grunt: `npm install -g grunt grunt-cli`
5. Install bower: `npm install -g bower`
6. Install yo: `npm install -g yo` --> complains that `Your Node.js version is outdated.` but no worries
7. Install generator-angular-fullstack@2 (note version 2, rather than latest): `npm install -g generator-angular-fullstack@2`
8. Follow along with rest of waypoint instructions.

**Comment by rtperson** [comment]
I see this is closed, but has a pull request been created to fix the instructions?

**Comment by danielpsf** [comment]
I get it working using ```rm -Rf * nvm install 0.12 && nvm use 0.12 && nvm alias default v0.12 npm install -g yo && echo "export NODE_PATH=$NODE_PATH:/home/ubuntu/.nvm/versions/node/v0.12.7/lib/node_modules" >> ~/.bashrc && source ~/.bashrc && npm install -g bower && npm install -g grunt-cli && npm install -g generator-angular-fullstack && yo angular-fullstack```

instead of what it says to do (http://freecodecamp.com/challenges/waypoint-get-set-for-basejumps)

**Comment by Hasco** [comment]
@mizujin Have u found solution for esprima issue? cause I have the same problem?
